### ISSUE TITLE
stack() casts to float when called on levels not requiring it

### ISSUE THREAD

**Issue by toobaz** [pandas-dev/pandas#17886]
#### Code Sample, a copy-pastable example if possible

```python
In [2]: df = pd.DataFrame(0, index=range(3),
   ...:                   columns=pd.MultiIndex.from_tuples([('w', 'a', 'j'),
   ...:                                                      ('w', 'a', 'k'),
   ...:                                                      ('w', 'b', 'j')],
   ...:                                                     names=['v', 'x', 'y']))

In [3]: df.stack(['x', 'y'])
Out[3]: 
v        w
  x y     
0 a j  0.0
    k  0.0
  b j  0.0
1 a j  0.0
    k  0.0
  b j  0.0
2 a j  0.0
    k  0.0
  b j  0.0
```

#### Problem description

Casting to float is not required (notice it would have been if ``df.stack()`` had been called without arguments).

#### Expected Output

```python
In [4]: df.stack(['x', 'y']).astype(int)
Out[4]: 
v      w
  x y   
0 a j  0
    k  0
  b j  0
1 a j  0
    k  0
  b j  0
2 a j  0
    k  0
  b j  0

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 6793a136dde139d968948b6ec9193cc3535abf6f
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.21.0rc1+13.g6793a136d.dirty
pytest: 3.0.6
pip: 9.0.1
setuptools: None
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 5.1.0.dev
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: None
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

**Comment by jreback** [comment]
xref #17845

**Comment by jreback** [comment]
again a special case where you have a cartesian product

**Comment by toobaz** [comment]
For future reference (I'll try to produce PRs for this and #17845 when I find the time): this can actually be catched without actually looking at index values, because every time we stack all levels _but one_ we obviously don't introduce any missing value.

**Comment by toobaz** [comment]
Other reminder: when this is fixed, remove [workaround](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/frame/test_reshape.py#L171) in tests.

**Comment by jreback** [comment]
@toobaz hopefully fixing this issue will *break* that test (until work-around is removed)!

**Comment by toobaz** [comment]
> @toobaz hopefully fixing this issue will break that test (until work-around is removed)!

Well, no, the workaround should just then cast from int... to int, but it should still work.

**Comment by rhshadrach** [comment]
With `future_stack=True`, the result is now integral. Closed by #55448
